docs/eeps/draft-class-loading.md

```diff
@@ -0,0 +1,524 @@
+        final Class<MavenArtifactFinder> clazz = loadImplClass();
+        try {
+            CONSTRUCTOR = clazz.getConstructor(Path.class);
+        } catch (final NoSuchMethodException ex) {
+            throw new LinkageError("Dependencies for Maven are not loaded correctly: " + CLASS_NAME, ex);
+        }
+    }
+
+    private static final Constructor<MavenArtifactFinder> CONSTRUCTOR;
+}
+```
+
+On the other hand, the concrete class `MavenArtifactFinderImpl` in `embulk-deps` actually calls Apache Maven ([Apache Maven Artifact Resolver](https://maven.apache.org/resolver/), which comes from [Eclipse Aether](https://maven.apache.org/aether.html)).
+
+```
+package org.embulk.deps.maven;
+
+import java.io.FileNotFoundException;
+// ...
+import java.nio.file.Path;
+// ...
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+// ...
+import org.eclipse.aether.resolution.ArtifactDescriptorException;
+import org.eclipse.aether.resolution.ArtifactDescriptorRequest;
+import org.eclipse.aether.resolution.ArtifactDescriptorResult;
+// ...
+import org.eclipse.aether.supplier.RepositorySystemSupplier;
+import org.embulk.plugin.MavenPluginType;
+import org.embulk.plugin.maven.MavenArtifactFinder;
+// ...
+import org.embulk.plugin.maven.MavenPluginPaths;
+
+public class MavenArtifactFinderImpl extends MavenArtifactFinder {
+    public MavenArtifactFinderImpl(final Path localMavenRepositoryPath) throws FileNotFoundException {
+        // ...
+    }
+
+    @Override
+    public final MavenPluginPaths findMavenPluginJarsWithDirectDependencies(final MavenPluginType pluginType, final String category)
+            throws FileNotFoundException {
+        // ...
+
+        final ArtifactDescriptorResult result;
+        try {
+            result = ...;
+        } catch (ArtifactDescriptorException ex) {
+            // ...
+        }
+
+        return ...;
+    }
+
+    // ...
+
+    private final RepositorySystemSupplier repositorySystemSupplier;
+    private final RepositorySystem repositorySystem;
+    private final RepositorySystemSession repositorySystemSession;
+}
+```
+
+The `embulk-core` and `embulk-deps` artifacts are tightly coupled by their nature. Different versions of them must not be paired. They are separated only to be loaded by different class loaders.
+
+Embulk's executable JAR file
+-----------------------------
+
+Almost all the Embulk users use Embulk's packaged executable JAR file while it is possible for users to embed `embulk-core` and `embulk-deps` by themselves. The next question is how to structure the executable JAR file.
+
+The hidden dependency libraries and `embulk-deps` cannot be extracted into the executable JAR file so that they are not loaded by the top-level class loader. Instead, we decided to include these library JAR files "as-is" nested in the executable JAR file. This is called the "JAR-in-JAR" technique, traditionally known for [One-JAR](https://one-jar.sourceforge.net/).
```

Comment:
What do you think about adding a built-in plugin explanation?

ex)
we decided to include these library
-> 
we decided to include these libraries(include built-in plugins)


IIUC, `*.jar` files contain (1) embulk-deps.jar itself, (2) hidden dependency libraries which load by embulk-deps (3) built-in plugins.

Reply:
Sounds reasonable. I added 

> (as well as the built-in plugins explained below)

in the sentense.

----

> IIUC, `*.jar` files contain (1) embulk-deps.jar itself, (2) hidden dependency libraries which load by embulk-deps (3) built-in plugins.

and (4) dependency libraries that are used from built-in plugins, too.

(More precisely, "hidden dependency libraries which load by embulk-deps" is not 100% correct. Thease hidden libraries are loaded "with" embulk-deps. embulk-deps does not have any class loading feature by itself.)